### ISSUE TITLE
Implement List Functionality for AST_Blocks

### ISSUE THREAD

**Issue by LapsededPacifist** [uchicago-cs/chiventure#1069]
Currently AST_blocks have prev and next pointers to other AST_blocks in order to be used as lists.
Functionality for adding, removing, and search with a list of AST_blocks needs to be added.
Note for adding to the list: The order that the actions appear in the list is often very important, so add needs to be able to add to a specific place in the list.

**Comment by jasulcaf** [comment]
Added `list_contains_AST_block` and `list_add_AST_block` functionality

- `list_contains_AST_block` takes in an AST_block_t (which has fields for a potential linked list) and block_type as arguments. It returns a boolean value on whether or not there exists an AST_block_t that is of the same block_type
- `list_add_AST_block` takes in an AST_block_t linked list, a newly made AST_block_t to add to linked list, and an integer that details what spot in the linked list we want to add the new AST_block_t in the linked list

Something to note is that we have also decided to completely get rid of the "prev" field in an AST_block_t. Essentially, there was no real reason as to why we would need this field over a regular single_linked_list. Not only this, but this 'prev' field was also causing memory Issues since AST_block_free is a recursive function (and as expected, if you try to free the current node, go to the next node, and try to free the previous node again, some nasty results are bound to happen).

**Comment by jasulcaf** [comment]
Added `list_how_many_AST_block` functionality

- `list_how_many_AST_block` takes in an AST_block_t (which has fields for a potential linked list) and returns how long the linked list is (minimum value of 1).

This function was added to primarily give more flexibility for the user when calling `list_add_AST_block.` Before, telling the function to add an AST_block_t in the 50th spot of the linked list when the linked list was only 3 long would have broken the code. Now, `list_add_AST_block` can take this scenario and work (it will simply add the specified AST_block_t as the "tail" of the list) which takes the previous burden of having to track how long the linked list is off the User. `list_how_many_AST_block` can also be useful if at any point in the future there is some type of restriction on how AST_block_t can be linked at any one time, or in general when users want to know how long the linked list is at the moment.

**Comment by jasulcaf** [comment]
Added `list_remove_AST_block` functionality

- `list_remove_AST_block` takes in an AST_block_t (which has fields for a potential linked list) as well as a block_type that specifies what we are removing. This function will remove all AST_block_t that have a matching block type from the linked list _and_ free them. Returns SUCCESS when there where items removed from the linked list, FAILURE when there was nothing to remove

**Comment by LapsededPacifist** [comment]
Updated searching so that it checks if the pointer to a block is already in the list.
This is used so that we can check if a block is already in the list before we try to add it. We realized that we would have a problem if we had a blocked duplicated in the list, as it would create an infinite loop. Since none of our functions could deal with that, we added the check to make sure it happens.

Also the indexing was changed to start at 0 instead of 1. This is just for consistency with the field of computer science.

**Comment by jasulcaf** [comment]
As a summary:

During the process of review in the Pull Request we decided to no longer implement `list_contains_AST_block` as well as `list_add_AST_block.` This is because currently we have no unique way of identifying AST_blocks without having to check if two pointers are equal; and there is currently no reason to worry about the specific location of AST_blocks in lists (though if in the future this is a feature that wants to be implemented, list_add_AST_block should be reconsidered). As such the overview of functionality that were added with this Issue are:

- `append_list_AST_block` : Function for adding a new AST_block to the end of a linked list
- `prepend_list_AST_block` : Function for adding a new AST_block to the start of a linked list
- `list_how_many_AST_block` : Function that checks how long an AST_block_t linked list is
- `list_remove_AST_block` : Function to remove a specific AST_block_t from a linked list (if it is found). This function _does_ free the removed AST_block_t
- `AST_free` : Function to free a _singular_ AST_block instead of an entire linked list. This was added during Pull Request Review and is essentially an abstraction of how AST_block_free worked (instead of having to free _all_ blocks, there now exists functions to free singular and another function to free all blocks; AST_free and AST_block_free respectively. For more information, go [here](https://github.com/uchicago-cs/chiventure/pull/1081#issuecomment-852526431)).
